### ISSUE TITLE
"Stricter" TypeScript

### ISSUE THREAD

**Issue by RyanCavanaugh** [Microsoft/TypeScript#274]
This a meta-bug for tracking a set of things that we could address with a compiler flag that tightens certain aspects of TypeScript that users generally perceive as too loose.

Will add to this list as appropriate
 * #228: Enforce expressions in `return` statements in type-annotated functions
 * #191: Don't allow mismatched constant string overloads to be selected due to other parameters
 * #21: Disallow use before definition (for types? for everything?)
 * ~~#360: Error when generic type inference produces *ex nihilo* `{}`~~ - now a default

Contentious issues that have been mentioned:

Done!
 * #391: Detecting accidental surplus properties

Not happening:
 * ~~#222: Function argument bivariance~~
   * We would want some strong motivating examples here; this behavior is desired in a lot of scenarios and is less "surprising" than others.
   * This isn't a plausible design path. See comments here https://github.com/Microsoft/TypeScript/issues/6102#issuecomment-164629334

**Note:** This is not a "fork the language" flag. The *type system itself* would be unchanged under this flag; it would simply change some operations from being allowed to being errors. For example, we would not change the order of overload resolution, or change the type of `null`, because that would have non-local effects and everyone would have to agree on whether or not the flag was on.

**Comment by danquirk** [comment]
Note #1740 as a potential additional error level, although it may be too much of the 'fork the language' case and there're better solutions if we do real 'this' typing in the core language.

**Comment by samwgoldman** [comment]
I think it would be really nice if I could ask the compiler to warn me when I do this:
```typescript
var data:Model = JSON.parse(json);
``` 

Instead of "silently" assigning the value of type `any` to `data:Model`, I would want the compiler to complain.

The compiler would stop complaining if I provided a type assertion. For example:
```typescript
var data:Model = <Model>JSON.parse(json);
```

I like this because I am assuming responsibility for the type cast. If `json` doesn't represent data that conforms to the `Model` interface, that's my fault.

Thoughts?

**Comment by RyanCavanaugh** [comment]
You can get the desired behavior by changing the return type of `JSON.parse` to `{}`. Disallowing all uses of `any` in its intended form seems far too strict.

**Comment by hesselink** [comment]
Since you're looking for motivating examples for disallowing function argument bivariance: we have code where we pass around constructor functions. These constructors accept an argument of an interface, but one of them incorrectly assumed it would get one specific concrete implementation of this interface instead. We expected this kind of thing to be caught but instead it was a runtime bug.

**Comment by markbook2** [comment]
It would be good to have ability to raise warning/error for all properties which don't have   AccessibilityModifier like `private | public | protected` in the stricter mode.

**Comment by afrische** [comment]
@markbook2 you can do that easily today with a regular expression( see below). 
Plus, this is a meta-bug, so you should probably file a new issue and reference this.
`````
<!-- URL that generated this code: -->
<!-- http://txt2re.com/index-javascript.php3?s=%20%20%20somename%20=%20somevalue;&5&3&7&-32&8&1&-31 -->

<html>
  <body>
    <script language=javascript>
      var txt='   somename = somevalue;';

      var re1='(\\s+)';	// White Space 1
      var re2='((?:[a-z][a-z]+))';	// Word 1
      var re3='(\\s+)';	// White Space 2
      var re4='(=)';	// Any Single Character 1
      var re5='(\\s+)';	// White Space 3
      var re6='((?:[a-z][a-z]+))';	// Word 2
      var re7='(;)';	// Any Single Character 2

      var p = new RegExp(re1+re2+re3+re4+re5+re6+re7,["i"]);
      var m = p.exec(txt);
      if (m != null)
      {
          var ws1=m[1];
          var word1=m[2];
          var ws2=m[3];
          var c1=m[4];
          var ws3=m[5];
          var word2=m[6];
          var c2=m[7];
          document.write("("+ws1.replace(/</,"&lt;")+")"+"("+word1.replace(/</,"&lt;")+")"+"("+ws2.replace(/</,"&lt;")+")"+"("+c1.replace(/</,"&lt;")+")"+"("+ws3.replace(/</,"&lt;")+")"+"("+word2.replace(/</,"&lt;")+")"+"("+c2.replace(/</,"&lt;")+")"+"\n");
      }
    </script>
  </body>
</html>
````

**Comment by adidahiya** [comment]
@markbook2 @afrische [TSLint has a rule](https://github.com/palantir/tslint/#supported-rules) for enforcing explicit visibility on class members / methods called `member-access`

**Comment by milesrout** [comment]
>I think --strict should include --noImplicitAny.

:+1:

**Comment by skogsbaer** [comment]
Here is another motivating examples for disallowing function argument bivariance:

In our typescript code base, we have an interface for classes having an `equals` method 

```typescript
export interface Eq {
    equals(x: any): boolean;
}
```

Because of function argument bivariance, classes can implement this interface "incorrectly", that is by narrowing the argument type `any`. For example:

```typescript
class C implements Eq {
    constructor(private c: number) {}
    // I would like to get an error here, stating that the method signature in 
    // the Eq interface does not match the implementation given here. 
    // In OO-languages such as C# or Java, you get this error.
    equals(that: C) {
        return this.getNum() === that.getNum();
    }
    getNum() {
        return this.c;
    }
}

class D {
    constructor(private d: number) {}
    equals(that: D) {
        return this.d === that.d;
    }
}
```

We now write a function that searches in an array of `Eq`-objects for an element:

```typescript
function findInArray(arr: Array<Eq>, x: Eq): number {
    for (let i = 0; i < arr.length; i++) {
        if (arr[i].equals(x)) {
            return i;
        }
    }
    return -1;
}
```

Now it's rather easy to trigger an unexpected runtime error:

```typescript
const arr = [new C(1), new C(2)];
const i = findInArray(arr, new D(2));
console.log(i);
```

The program now aborts with `TypeError: undefined is not a function`

**Comment by mhegazy** [comment]
We have moved to a more piecemeal approach to such checks, e.g. control flow checks, strict this, strict null, etc.. feels like this issue is obsolete. @RyanCavanaugh any objections to closing it?

**Comment by magnushiie** [comment]
Another case where bivariant arguments really hurt is the React and Redux world, where mismatches in React component props are one of the most frequent source of errors.

Currently it's perfectly valid to have:
```
interface MyComponentProps {
  title: string;
}
export class MyComponent extends React.Component<MyComponentProps, void> {
  // code that needs this.props.title
}
const MyComponentAlias: React.ComponentClass<{}> = MyComponent;
// use of MyComponentAlias blows up, as it doesn't require title, but title is needed for MyComponent
```

The ```const``` in the above example is not a common use case, but I'm currently trying to make react-redux (especially the connect function) typings stricter, so props interface changes would highlight other changes that need to be made, but the bivariance only allows checking that the property field types are right, not that all properties are present where they need to be (and therefore also allows typos).

**Comment by DemiMarie** [comment]
I would like a flag under which TypeScript would reject any program that it could not prove to be free of type errors at runtime.

In addition to proving that a large class of bugs (type errors) are impossible, an AOT compiler (for servers) for "sound" typescript could erase the types at runtime for a perf gain.

**Comment by RyanCavanaugh** [comment]
@DemiMarie what you're describing is not broadly possible given the constraints of our type system. As one example of many, we have no plausible way to prevent mutation after aliasing a structure with a less-specific type introducing an incorrect value into that structure, e.g. :
```ts
const a = [1, 2, 3];
const b: Array<string | number> = a;
b[0] = 'oops';
console.log(a[0] + 1); // 'oops1'
```

We've done a good job implementing almost everything in the OP so I'm going to close this due to a lack of usefulness of a meta-issue. Some good discussion happening at #9642 about how to get these stricter options on by default in places where it wouldn't be troublesome.
